drivers/mailbox/bflb-ipc.c

```diff
@@ -157,79 +262,127 @@ static const struct irq_domain_ops bflb_ipc_irq_ops = {
 	.xlate = bflb_ipc_domain_xlate,
 };
 
+#if 0
+/* Shouldn't actually be fail as we clear the High/Low registers in a EOI
+ * but this protects if we screw up our mailbox handling
+ */
+static bool bflb_ipc_mbox_can_send(struct mbox_chan *chan)
+{
+	struct bflb_ipc *ipc = to_bflb_ipc(chan->mbox);
+
+	/* check the low register first as we clear that last in our EOI, so this
+	 * should protected to a limited extent
+	 */
+	u32 mbox_low = readl(ipc->base[2] + IPC_REG_ILSLR);
+	u32 mbox_high = readl(ipc->base[2] + IPC_REG_ILSHR);
+
+
+	if (mbox_low | mbox_high)
+		dev_warn_ratelimited(ipc->dev, "%s: low: 0x%08x high: 0x%08x\r\n", __func__, mbox_low, mbox_high);
+
+	writel(0, ipc->base[2] + IPC_REG_ILSLR);
+	writel(0, ipc->base[2] + IPC_REG_ILSHR);
+
+	return !(mbox_low | mbox_high);
+}
+#endif
+
 static int bflb_ipc_mbox_send_data(struct mbox_chan *chan, void *data)
 {
 	struct bflb_ipc *ipc = to_bflb_ipc(chan->mbox);
 	struct bflb_ipc_chan_info *mchan = chan->con_priv;
-	u32 hwirq;
+	struct bflb_mbox_msg *msg = data;
+	u32 tmpVal = (mchan->service_id << 16) | (mchan->op_id & 0xFFFF);
 
-	hwirq = bflb_ipc_get_hwirq(mchan->client_id, mchan->signal_id);
+#if 0
+	if (!bflb_ipc_mbox_can_send(chan))
+		return -EBUSY;
+#endif
 
-	dev_dbg(ipc->dev, "%s: hwirq: %u\n", __func__, hwirq);
+	dev_dbg(ipc->dev, "%s %d: cpu: %d singal: %d op: %d (0x%x) param: %d", __func__, msg->id, mchan->cpu_id, mchan->service_id, mchan->op_id, tmpVal, msg->param);
 
-//	writel(hwirq, ipc->base + IPC_REG_SEND_ID);
+	// /* write our signal number to high register */
+	writel(tmpVal, ipc->base[2] + IPC_REG_ILSHR);
```

Comment:
Since all mbox channels share the same register, there needs to be some locking here if we want to support multiple mbox channels